server/src/main/java/org/apache/druid/server/http/DataSourcesResource.java

```diff
@@ -700,4 +704,85 @@ static boolean isSegmentLoaded(Iterable<ImmutableSegmentLoadInfo> serverView, Se
     }
     return false;
   }
+
+  @PUT
+  @Path("/{dataSourceName}/markUsed")
+  @Produces(MediaType.APPLICATION_JSON)
+  @ResourceFilters(DatasourceResourceFilter.class)
+  public Response enableDatasourceSegments(
+      @PathParam("dataSourceName") String dataSourceName,
+      MarkDatasourceSegmentsPayload payload
+  )
+  {
+    if (payload == null || !payload.isValid()) {
+      return Response.status(Response.Status.BAD_REQUEST).entity(
+          ImmutableMap.of(
+              "error",
+              "Either a valid interval or segmentIds array must be specified"
+          )
+      ).build();
+    }
+
+    final ImmutableDruidDataSource dataSource = getDataSource(dataSourceName);
+    if (dataSource == null) {
+      return Response.noContent().build();
+    }
+
+    boolean success;
+    try {
+      if (payload.getInterval() != null) {
+        success = databaseSegmentManager.enableSegments(dataSource.getName(), payload.getInterval());
+      } else {
+        success = databaseSegmentManager.enableSegments(dataSource.getName(), payload.getSegmentIds());
```

Comment:
For the segmentId case, I think it would be informative if the response gave per-segment success/fail info

Reply:
It now returns a JSON object with the count of modified segments.

Reply:
It no longer returns anything but the status code!